### DISCUSSION TITLE
httpx middleware

### DISCUSSION THREAD

**Discussion by clarkreiz** [encode/httpx#3392] [Q&A] [answered]
I want to use httpx along with my custom middleware. Currently, I need middleware for logging (after receiving a response) and authentication (after sending a request). In the future, I might also need one for SSL.

I realized that I can’t use event_hooks because they don’t allow me to add anything to the headers. Is there an approach similar to [request-middleware](https://github.com/jmcarp/requests-middleware) that would work for this?

**Comment by lovelydinosaur** [comment, marked as answer]
Try taking a look at custom transports, perhaps?
They'll let you do pretty much any customisation of the request-response cycle.

There's a similar discussion here... https://github.com/encode/httpx/discussions/3073
